pkg/pillar/devicenetwork/dns.go

```diff
@@ -150,21 +154,27 @@ func ResolveWithPortsLambda(domain string,
 						errs = append(errs, err)
 					}
 					if response != nil {
-						resolvedIPsChan <- response
+						select {
+						case <-quit:
+						case resolvedIPsChan <- response:
+						}
 					}
-					<-work
-					wg.Done()
 				}(dnsIPCopy, srcIPCopy)
 			}
 		}
 	}
 
 	wgChan := make(chan struct{})
+
 	go func() {
 		wg.Wait()
 		close(wgChan)
 	}()
 
+	defer func() {
+		close(quit)
+		<-wgChan
+	}()
 	select {
```

Comment:
Okay, one thing I don't quite get is if we want any resolved IP addr, or all resolved IP addrs. Seems to me that we return first successful resolve. Because if we were to send all of the possible options, wrapping wg.Wait() in goroutine wouldn't make sense, we do wg.Wait() in function, block till we check all the options and then deal with results, because select is also blocking. 

In any of case I'd suggest we extract lookup function (that two loops with DNS and IPs) and write something like this:


```go

errs := []error
resolvedIPCh, errCh, outOfOptions := lookupIP(/* ... */)
select {
  case ip := <- resolvedIP:
     return ip, nil
  case <-outOfOptions:
     ips, err := net.LookupIP(domain)
     if err != nil {
     }
     // ...
  case err := <- errCh:
     errs.append(err)

} 

```

I think this way it's way more straightforward, I can write full snippet if you think it's a good approach @christoph-zededa

Reply:
I am curious what that would look like.
I think breaking it up, would make it more readable.

Reply:
Hold my Apfelschorle...